include/envoy/buffer/buffer.h

```diff
@@ -194,6 +197,152 @@ class Instance {
+
+    const auto extension =
+        std::is_signed<T>::value && Size < sizeof(T) && bytes[most_significant_read_byte] < 0
+            ? static_cast<T>(static_cast<typename std::make_unsigned<T>::type>(all_bits_enabled)
+                             << (Size * CHAR_BIT))
+            : static_cast<T>(0);
+
+    return fromEndianness<Endianness>(static_cast<T>(result)) | extension;
+  }
+
+  /**
+   * Copy a little endian integer out of the buffer.
+   * @param start supplies the buffer index to start copying from.
+   * @param Size how many bytes to read out of the buffer.
+   */
+  template <typename T, size_t Size = sizeof(T)> T peekLEInt(uint64_t start = 0) {
+    return peekInt<T, ByteOrder::LittleEndian, Size>(start);
+  }
+
+  /**
+   * Copy a big endian integer out of the buffer.
+   * @param start supplies the buffer index to start copying from.
+   * @param Size how many bytes to read out of the buffer.
+   */
+  template <typename T, size_t Size = sizeof(T)> T peekBEInt(uint64_t start = 0) {
+    return peekInt<T, ByteOrder::BigEndian, Size>(start);
+  }
+
+  /**
+   * Copy an integer out of the buffer and drain the read data.
+   * @param Size how many bytes to read out of the buffer.
+   * @param Endianness specifies the byte order to use when decoding the integer.
+   */
+  template <typename T, ByteOrder Endianness = ByteOrder::Host, size_t Size = sizeof(T)>
+  T drainInt() {
+    auto const result = peekInt<T, Endianness, Size>();
+    drain(Size);
+    return result;
+  }
+
+  /**
+   * Copy a little endian integer out of the buffer and drain the read data.
+   * @param Size how many bytes to read out of the buffer.
+   */
+  template <typename T, size_t Size = sizeof(T)> T drainLEInt() {
+    return drainInt<T, ByteOrder::LittleEndian, Size>();
+  }
+
+  /**
+   * Copy a big endian integer out of the buffer and drain the read data.
+   * @param Size how many bytes to read out of the buffer.
+   */
+  template <typename T, size_t Size = sizeof(T)> T drainBEInt() {
+    return drainInt<T, ByteOrder::BigEndian, Size>();
+  }
+
+  /**
+   * Copy a byte into the buffer.
+   * @param value supplies the byte to copy into the buffer.
+   */
+  void writeByte(uint8_t value) { add(std::addressof(value), 1); }
+
+  /**
+   * Copy value as a byte into the buffer.
+   * @param value supplies the byte to copy into the buffer.
+   */
+  template <typename T> void writeByte(T value) { writeByte(static_cast<uint8_t>(value)); }
+
+  /**
+   * Copy an integer into the buffer.
+   * @param value supplies the integer to copy into the buffer.
+   * @param Size how many bytes to write from the requested integer.
+   * @param Endianness specifies the byte order to use when encoding the integer.
+   */
+  template <ByteOrder Endianness = ByteOrder::Host, typename T, size_t Size = sizeof(T)>
+  void writeInt(T value) {
+    static_assert(Size <= sizeof(T), "requested size is bigger than integer being read");
```

Comment:
I find this error a bit confusing: should it say being written?

Reply:
you're right